api/query/query.go

```diff
@@ -17,6 +20,244 @@ import (
 	"github.com/web-platform-tests/wpt.fyi/shared"
 )
 
+var browsers = []string{
+	"chrome",
+	"edge",
+	"firefox",
+	"safari",
+}
+
+// TODO: Establish query execution plans over internal data schema.
+type plan interface{}
+
+type query interface {
+	toPlan([]int64) plan
+}
+
+type runQuery struct {
+	runIDs []int64
+	query
+}
+
+func (rq runQuery) toPlan() plan {
+	return rq.query.toPlan(rq.runIDs)
+}
+
+type testNamePattern struct {
+	pattern string
+}
+
+func (tnp testNamePattern) toPlan(runIDs []int64) plan {
+	return nil
+}
+
+type testStatusConstraint struct {
+	browserName string
+	status      int64
+}
+
+func (tsc testStatusConstraint) toPlan(runIDs []int64) plan {
+	return nil
+}
+
+type not struct {
+	not query
```

Comment:
`not{not: testNamePattern{pattern: "abc"}}` not not readable

I think we'll have to ignore lint saying we should name the params in Golang, and just write
`not{testNamePattern{"abc"}}`